.github/workflows/tests.yaml

```diff
@@ -4,69 +4,130 @@ on:
   pull_request:
     types: [opened, synchronize]
     paths:
-        - 'api/**'
-        - 'managers/**'
-        - 'mappings/**'
-        - 'model/**'
-        - 'processes/**'
-        - 'services/**'
-        - 'tests/**'
-        - 'utils/**'
-        - 'dev-requirements.txt'
-        - 'requirements.txt'
-        - 'main.py'
-        - 'Makefile' 
+      - "api/**"
+      - "managers/**"
+      - "mappings/**"
+      - "model/**"
+      - "processes/**"
+      - "services/**"
+      - "tests/**"
+      - "utils/**"
+      - "dev-requirements.txt"
+      - "requirements.txt"
+      - "main.py"
+      - "Makefile"
 
 jobs:
-  test:
+  docker-setup:
+    name: Start Docker Containers
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Start Docker containers
+        run: docker compose up -d
+      - name: Wait for services to be ready
+        run: sleep 90
+
+  unit-tests:
+    name: Unit Tests
     runs-on: ubuntu-latest
     env:
       IS_CI: "true"
       AWS_ACCESS: ${{ secrets.AWS_ACCESS_KEY_ID }}
       AWS_SECRET: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
       ENVIRONMENT: local
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v4
       - name: Set up Python 3.9
-        uses: actions/setup-python@v2
+        uses: actions/setup-python@v5
         with:
           python-version: "3.9"
       - name: Cache dependencies
         uses: actions/cache@v4
         with:
           path: ~/.cache/pip
-          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
+          key: ${{ runner.os }}-pip-${{ hashFiles('requirements*.txt') }}
           restore-keys: |
             ${{ runner.os }}-pip-
             ${{ runner.os }}-
-      - name: Install GCC 12
-        run: |
-          sudo apt update
-          sudo apt install -y gcc-12 g++-12
-          sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 60
-          sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-12 60
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip wheel
           pip install -r dev-requirements.txt
           pip install -r requirements.txt
-      - name: Run unit tests
-        run: |
-          make unit
-      - name: Start Docker containers
-        run: |
-          docker compose up -d
-      - name: Wait for docker containers to be ready
-        run: |
-          echo "Waiting for docker container..."
-          sleep 90
-      - name: Run functional tests
+          pip install pytest-xdist
+      - name: Run Unit Tests
+        run: pytest tests/unit -n 4 -v
+
+  functional-tests:
```

Comment:
I recommend not having separate jobs for these. While it's nice to be able to run them at the same time, it mean we have to checkout, cache, and install the dependencies all over again.